### ISSUE TITLE
Discussion: Cleaning up menus

### ISSUE THREAD

**Issue by manongjohn** [opentoonz/opentoonz#2524]
There has been some discussion on Discord about how the current state of our menu's organization can be improved upon.

I'm opening this issue to discuss how best to reorganize the main menus so they make sense to new users without being overwhelming. Eventually I'll turn it into a PR.

**Comment by gab3d** [comment]
Could you please post a link to that Discord thread? So to see what's been discussing there?

Right off the bat I'd say that the File menu is a little overpopulated, and also perhaps that I often find myself looking for some option between the Level, Xsheet and Cell menus, without knowing exactly why it is there, and not in any other menu. :roll_eyes:

**Comment by RodneyBaker** [comment]
Gabriel,
Not much was discussed via discord beyond the desire to discuss menus and how they can be improved.  As such the discussion is right here... with Manongjohn taking on the role as primary developer.

Much of the file menu can be adjusted by users but the current organization leaves something to be desired as there isn't much of a logical flow to the order and placement of menu listings.  

In a few places the menus might be best served by simply being alphabetized.
This might especially be true of the Window listing.
It seems to start off alphabetized but then loses that about 3/4 of the way down.
The listing under File could use a bit of reorganization to collect like items such as Open, Import, Load, Convert (processes dealing mostly with ingest of data/files.   Similarly processes such as Save, Export, Print are similar in that they focus on data output.  One approach might be to create submenus but a concern therein might be how that might impact users that use keystrokes to drill down and launch a command (example:  File > Save All would be a two key combo in this regard whereas placing the Save commands one more menu level down might make it a three key process)  Of course the shortcut keys would still be available as is now.

Bottom line here:  Submenus (if used) would assist with the overpopulation of the File menu.

In the reorganization there are also a few menu commands that can be added that are not currently in place.

**Comment by beeheemooth** [comment]
I noticed that many users are looking for options: copy / paste / delete in the right-click menu, but they are not there. In many programs, these operations are available in the RB menu.

**Comment by artisteacher** [comment]
I think having submenus in some strategic places would make the menu a lot easier to work with overall. While I see the downside of extra stylus clicks or etc, some parts of the menu are so overpopulated that it’s pretty necessary.  Alphabetizing is also a good idea. 

Starting with the File menu, making submenus for actions of the same type seems to be the way to go (ie. Save All, Save Level, Save Scene all together). 

I agree with @gab3d that it’s not always obvious right away why certain options are under cell vs Xsheet or etc.

**Comment by DarrenTAnims** [comment]
Hi all. I've edited my menu quite a lot and this is my current configuration. I've attached a gif showing the menus and my menu file. If you want to try this out yourself, rename the xml file from **room1_menubar.txt** to **room1_menubar.xml**
and drop the xml file into your Rooms folder. For a 1.3 release, this is at:
C:\OpenToonz stuff\profiles\layouts\personal\Default.YourName\

This xml works just for one room though. There's a different menu layout and xml file for each room. You'd need to copy it into the other room's xml files to use this menu for their rooms.

The main things you'll notice are the File menu has been rationalised, grouping similar items. I've moved the prefs and shortcuts items out to the new Tools menu and added some items there from the View menu. I've created a new View menu and placed it next to Edit, where it normally is on other apps and that has a shorter list of items that can be viewed. I've moved the "checks" checkboxes into their own menu. Finally, I've grouped the items in the Windows menu into similar groups.

There's other menus I've not done much with and this is partially because the items are in the right place (but could probably be better grouped) or I don't use them enough to have an opinion yet

![darrent menus](https://user-images.githubusercontent.com/28575907/52955492-b06ace00-3384-11e9-89ac-27d3c628c330.gif)

[room1_menubar.txt](https://github.com/opentoonz/opentoonz/files/2875912/room1_menubar.txt)

**Comment by RodneyBaker** [comment]
I really like that direction @DarrenTAnims.
The New, Load  and Save submenus go a long way in organizing the flow of operations and reducing the clutter.

**Comment by konero** [comment]
I wouldn't put save functions behind a sub-menu as technically they're not related and are destructors. All file operations should be listed under the File menu, any function that can load, duplicate, export or overwrite the opened file.

Anything that's equivalent to "save copy" or "save duplicate" such as export options would be OK under a sub-menu, such as a list of different filetypes.

The "open" operations should be near the top of their counterparts as they're a launch function for most people. The conventional wisdom is usually New, Open, Recent, Save, As, Export, Close, Quit.

IMHO you're breaking the mould here a bit. If the menus must be reorgranized, copy directly from Clip Studio Paint since it's the software most often paired with.

**Comment by artisteacher** [comment]
I wish I could test out that menu, but it doesn't seem to work on macOS. However, even just consolidating the load and new options into sub-menus would really help the File menu. 

@konero - what does the File menu look like in Clip Studio Paint?

If reorganization happens, a link to the manual should really be added in the help menu. There's been a good amount of work to make it more up-to-date.

**Comment by gab3d** [comment]
@artisteacher 

> If reorganization happens, a link to the manual should really be added in the help menu. There's been a good amount of work to make it more up-to-date.

I strongly second that idea. 
Currently it must be the less-helpful "Help" menu out there! 😉 
And it's also true that nowadays the manual is much more up to date than it used to be.

**Comment by konero** [comment]
@artisteacher Not much different than OpenToonz to be fair.

What makes Harmony/Flash/Photoshop so well used is that they've remained the same for many years, so a tutorial, course or book from 2004 is still relevant today. It's also good to use standard computing practices so as not to keep introducing learning curves.

OpenToonz also has a book manual written in Japanese you can buy here: https://www.amazon.co.jp/dp/4777520617/ref=cm_sw_em_r_mt_dp_U_8N6CCb7SH7E2D

...so it might be unkind to change too much to the author, which I think is [Yoichi Senzui](https://twitter.com/sen_cha?lang=en).

Here's a layout I'd recommend, while trying to take in consideration the legacy users as well as new users.

> **Download: [room8_menubar.zip](https://github.com/opentoonz/opentoonz/files/2899974/room8_menubar.zip)**
> 
> - _**How to use:** create a new room and give it a name, navigate to your Toonz profile folder eg "C:\Toonz\profiles\layouts\personal\StudioGhibli.UserName" and then replace the menu xml file giving it the same name, reload OpenToonz._

I'd suggest moving "Clear Recent Level File List" and embed it inside the "Open Recent Level File" menu item, at the bottom with a separator above it as is commonly found in most software and also do the same for the Scene File items, but this can't be done by editing the profile files, it will save a little space.

Also by the way, the default room needs updating, when you create a new room a lot of menu items are missing, you have to add them manually each time. I've added every item to my menu file above so if anyone updates this stuff, be sure to update the default room too.

![optnz_menu](https://user-images.githubusercontent.com/19820721/53331144-18bb3180-38e8-11e9-83d8-4daddd5d9ad2.gif)

**Comment by artisteacher** [comment]
I don't think we can customize the menu on macOS 😕, so I haven't run into that problem with new rooms.

It's definitely important to follow standards and keep legacy users in mind. However, I hope we can do that without limiting ourselves to a design that doesn't work. On my small work laptop, I have to scroll down to see all of the file menu - that's just way too long. There are so many menu items that it can be hard to find the right item. Any changes will hopefully make menu items easier to find, which should benefit both legacy and new users.

In @konero's suggestion, I really like the "convert" submenu under "Levels", the "playback" submenu under Xsheet and the startup pop-up moved under "Windows".

While it may not be standard, I do think @DarrenTAnims suggestion of the "New", "Load" and "Save" submenus is very clear and makes it easier/faster to find those options. It also saves a ton of space, which I really appreciate. I'd vote for those submenus myself, but if they are not a possibility, does that overall organization still make more sense?  Organizing by action instead of by type?

- New Project
- New Scene
- New Level
---------------
- Load Scene
- Load Level
- Load Folder
- Load as Sub-Xsheet
- Load Color Model
--------------
- Save All
- Save Scene
- Save Level
- & etc...

**Comment by konero** [comment]
@artisteacher You can't modify it on Mac? Nothing happens if you right-click on a room tab and click "Customize Menu"? That'd be a bit unfair.

I see what you mean by the long list for File menu, I've had a harder look, here's some adjustments I made.

- Remove **Import Magpie File**: The software is dead.
- Put **Project** commands in their own sub-group.
- Put loading commands into a **Import / Load** sub-group.
- Put exporting commands into a **Export / Render** sub-group.
- Put scripting commands into a **Script** sub-group.

Suggestions:
- Move "**Touch Gesture Control**" into the Preferences window and remove it from the menu items.

Is this any better?

> **Download:** [room7_menubar.zip](https://github.com/opentoonz/opentoonz/files/2909927/room7_menubar.zip)

One caveat of sub-groups is can they be translated?

![ja_menu_tz](https://user-images.githubusercontent.com/19820721/53486103-bf860600-3a7f-11e9-88f1-f8a40d0799b8.jpg)

**Comment by artisteacher** [comment]
There’s not even a customize menu option if you right-click! There’s a context menu with some sub-menus, so they ought to be translatable. 

That length definitely looks better.

**Comment by manongjohn** [comment]
@artisteacher I noticed that to get to add new room and customize the menubar, you have to right-click over a room tab since each room tab's menubar is configured separately. Did you try there?

**Comment by artisteacher** [comment]
@manongjohn - I only get options to make a new room, or delete a room - nothing to customize the menu. Also, replacing the various menu files in the Stuff folder seems to have no impact.

**Comment by manongjohn** [comment]
I just looked at the code to verify. The customize menu bar option is only available for Windows builds and the menus are fixed based on what's hard coded in `createFullMenuBar()`.  Xmls are ignored. :(

**Comment by gab3d** [comment]
@DarrenTAnims 
In general, regarding the File menu, I see the idea of grouping related commands under a submenu as a good one for saving space. 
I can't decide yet if I like more the 
**Scene** / **Level** / **Project** / **Output** (which can also include Preview options) / **etc** 
or
**New** / **Load** / **Save** / **Output** / **etc** 
ones...

The first grouping would be more or less just making submenus out of the items that are between the current *separator lines* of the File menu.
The second option would imply movng more things around, naturally and a bit of rethinking of some of the logic behind the placement of some options in one or other menu.

I see no clear advantage of separating the **Check** menu options from the **View** one, because essentially the checking options affect directly the viewer and they seem to be placed in a logical site already. 
What I do feel it would be helpful is if the user could check/uncheck several of these options without having to reopen the menu again and again for each click. That behaviour is not very "user friendly" (and isn't either solved by separating them into a new Check menu I'm affraid). If the user could Shift Click the options to have them toggle without leaving the menu it'd be just great!

I do like very much the grouping you've done in the Windows menu, using separator lines.
I think it has two clear advantages:
- it allows to do a quick eyeball search by type of window.
- it eliminates a current problem with the alphabetical order and the translations to other languages (where this alphabetical order is broken, and options have to be found just by reading them one by one)
Thumbs up for that one!

I'm not so sure about the **Tools** menu name because in OT a tool is like the Brush tool, Animate tool, etc. and neither of the options there can be properly be called "tools". Having said that, I do feel some of the things you placed there deserve a cleaner home in the menus, like the Shift and Trace options, Preferences, Scripting commands, etc.  Lastly, I think the **Visualize vector as raster** and **Link Flipbooks** options are more logically placed where they are now (under the View menu) as they directly affect how the user interacts with the viewer.

For things like **Preferences**, **Configure Shortcuts** ,  I personally think they should belond to the **Edit** menu, as more and more software are putting their preferences options there, and users firstly look for them there too. Perhaps some other settings related options could be placed in the Edit menu too, like the Scene settings...

I can't decide where to put the **Shift and Trace** options, if moved from its current place, but I do feel in there they are somewhat hidden to the user, and being the great utility that it in fact is, we should bring it to a more prominent place in the interface.

With respect to the current **Level**, **Xsheet** and **Cells** menus, (contrary to my initial instinct, and) after having analysed why each option is placed in each menu, I've come to the conclusion that there is a very sound thinking regarding why they are made like this, and I think we should tend not to mess too much with them.

I do feel inconsistent that if the **Apply Lipsync Data to Column** is placed in the Xsheet menu, the **Import Magpie File** option is in the File menu, while esentially both do practically the same thing (perhaps in a different order, in one case the options window is opened before you load the file, and in the other one it asks for the file, and then it opens the options window). All in all, we should put these two on besides the other, wherever that happens to be the best place.

**Comment by artisteacher** [comment]
It certainly would be nice for macOS to have that capability at some point.

Here's what @konero's layout would look like on MacOS:
![screen shot 2019-02-27 at 2 59 29 pm](https://user-images.githubusercontent.com/21159116/53519049-5cf13200-3aa0-11e9-8242-8656a9b0795d.png)
Definitely a lot more manageable.

Question - why do we use the term "open" just for recent files?

**Comment by VidelXi** [comment]
Hi guys, I really don't like the idea of grouping save items into flyouts, it's not good usability.

The one posted by Konoro [here](https://github.com/opentoonz/opentoonz/issues/2524#issuecomment-467822223) is fine, just add export soundtrack to the export menu and the magpie option to the load menu and it's good. 🤷‍♂️ 

Here's the file menu in English on Windows:

![7896786](https://user-images.githubusercontent.com/22757323/53519805-07327f00-3acc-11e9-8909-c7e63c0eb834.jpg)

**Comment by konero** [comment]
Here's one last attempt from me. Try keep industrial users in mind.

> **Download: [room7_menubar.zip](https://github.com/opentoonz/opentoonz/files/2915729/room7_menubar.zip)**

![tz_menu_30](https://user-images.githubusercontent.com/19820721/53579071-cc365700-3b70-11e9-8f1d-ce71b5038887.gif)

**Comment by artisteacher** [comment]
That looks really good! 
- Personally I might move the proposed "Play" menu to between "Cells" and View"
- I think moving the various level commands from "File" to under the "Level" menu works well & might be more standard (similar to how Photoshop's layer menu works)
- The "Optimize", "Tonal Correction", and "Convert" submenus under the "Level" menu are great
- For the "Cells" menu, I wonder if it should be organized more like the cell context menu (with "Step" & "Each" as submenus)
- The organization of the "Windows" menu makes more sense

In the "Edit" menu, should the forward/backward options be moved under an "Arrange" submenu? That's pretty standard...

**Comment by konero** [comment]
@artisteacher All good points, here's a modified menu based on those.

> **Download: [room7_menubar.zip](https://github.com/opentoonz/opentoonz/files/2917520/room7_menubar.zip)**

![tz_menu_40](https://user-images.githubusercontent.com/19820721/53609608-9ff5f700-3bbf-11e9-8ebc-16fe28fe98e1.gif)

I added a render menu as an experiment, my colleague has it set up like this identical to the Xsheet Ghibli room since he's a compositor/sfx artist and does all the render stuff, he say's he prefers to have the output render and preview commands in the same place. There's also a command that's exclusive to the Ghibli room so I copied it over from the xml, I don't think it's available from the menu customize screen (Save Previewed Frames), at least I couldn't find it.
